adapter-jdbc/src/main/java/io/nosqlbench/adapter/jdbc/optypes/JDBCDMLOp.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2023 nosqlbench
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.nosqlbench.adapter.jdbc.optypes;
+
+import io.nosqlbench.adapter.jdbc.JDBCSpace;
+import io.nosqlbench.adapter.jdbc.exceptions.JDBCAdapterInvalidParamException;
+import io.nosqlbench.adapter.jdbc.exceptions.JDBCAdapterUnexpectedException;
+import io.nosqlbench.adapter.jdbc.exceptions.JDBCPgVectorException;
+import io.nosqlbench.adapter.jdbc.utils.JDBCPgVector;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.List;
+
+public abstract class JDBCDMLOp extends JDBCOp {
+    private static final Logger LOGGER = LogManager.getLogger(JDBCDMLOp.class);
+    protected final boolean isReadStmt;
+    // If the passed-in statement string has placeholder '?', it is a prepared statement.
+    protected final boolean isPreparedStmt;
+    protected final String pStmtSqlStr;
+    protected final List<Object> pStmtValList;
+
+    protected static ThreadLocal<Statement> jdbcStmtTL = ThreadLocal.withInitial(() -> null);
+
+    public JDBCDMLOp(JDBCSpace jdbcSpace,
+                     boolean isReadStmt,
+                     String pStmtSqlStr,
+                     List<Object> pStmtValList) {
+        super(jdbcSpace);
+        assert(StringUtils.isNotBlank(pStmtSqlStr));
+
+        this.isReadStmt = isReadStmt;
+        this.pStmtSqlStr = pStmtSqlStr;
+        this.pStmtValList = pStmtValList;
+        this.isPreparedStmt = StringUtils.contains(pStmtSqlStr, "?");
+
+        // NOTE:
+        // - The write DML statement MUST use prepared statement!
+        // - The read DML statement can use either a prepared or a regular statement.
+        if (!isReadStmt && !isPreparedStmt) {
+            throw new JDBCAdapterInvalidParamException(
+                "Write DML statement must use prepared statement format!");
```

Comment:
This should be an option for the user. For example, look at the cqld4.md docs on statement forms.

Reply:
I'm going to add this in a TBD list for future work